### ISSUE TITLE
Tensorflow Object Detection API using Faster RCNN on Android devices. 

### ISSUE THREAD

**Issue by AliceDinh** [tensorflow/models#4848]
Please go to Stack Overflow for help and support:

http://stackoverflow.com/questions/tagged/tensorflow

Also, please understand that many of the models included in this repository are experimental and research-style code. If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **What is the top-level directory of the model you are using**: object_detection
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Window 10
- **TensorFlow installed from (source or binary)**: binary, installed from pip: pip install tensorflow
- **TensorFlow version (use command below)**: 1.8
- **Bazel version (if compiling from source)**: None
- **CUDA/cuDNN version**: None
- **GPU model and memory**: None
- **Exact command to reproduce**: None

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
I used **TensorFlow Android Camera Demo**, which is **TF Detect**, at this link:
(https://github.com/tensorflow/tensorflow/tree/r1.9/tensorflow/examples/android)
I trained my own model using **SSD w/Mobilenet detector** using my own data, then I imported the frozen model (about **20MB**) to Android project and it was working well on my phone: **Samsung S7Edge, 4GB Ram**.
Now, I changed from SSD w/Mobilenet detector to **Faster RCNN** because Faster RCNN gives higher accuracy as many people stated. The frozen model is checked on Jupyter Notebook and the result is good. The frozen model is about **110MB** and I imported to Android project. With 4GB Ram of Samsung S7Edge I believe it will be okies but there is **no output result and the app crashed**. So, my questions are:

**Is Faster RCNN model not suitable for mobile? what should I do to make it work on Mobile?**

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by karmel** [comment]
@derekjchow - can you comment on device compatibility for Faster RCNN?

**Comment by derekjchow** [comment]
Faster-RCNN models tend to be very compute intensive (compared to SSD models). Additionally, Faster-RCNN models contain a crop-and-resize step that is not supported in the TFLite run time. For these reasons, we would not recommend using them for mobile inference.

We highly recommend using SSD models for mobile. In particular, we've done a lot of work to support 8-bit quantized inference for SSD models in [TFLite](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tensorflowlite.md). In addition, if you use TF-Lite, certain phones will support hardware acceleration via NNAPI.

Could you describe which SSD models you're working with? It's quite likely that you can get a more accurate SSD model. For example, you can try changing the feature extractor or input resolution. See our paper [Speed/accuracy trade-offs for modern convolutional object detectors](https://arxiv.org/pdf/1611.10012.pdf) for more details. In addition, we have also released FPN/PPN SSD networks which may also give additional performance.

**Comment by AliceDinh** [comment]
@derekjchow Thanks for your reply.
I am using SSD (ssd_mobilenet_v1_coco.config) to train my own model: detect numbers on ATM card (like SamsungPay app). The ATM card has the standard size 660x416 and digits on cards can be considered medium/small objects that challenging SSD detector.

If I use default SSD configuration, inputs are 660x416 images, annotations are made based on those inputs, after 18K steps, the loss is about 1.2 and the accuracy is low. 
![1](https://user-images.githubusercontent.com/20389638/43444442-ecd9b820-94cd-11e8-93e6-e4920182e398.png)

I reckoned that due to the resizing step of SSD to 300x300, my inputs' size 660x416 (width:height !=1:1) made the training time longer and the accuracy lower, so I changed my inputs to 660x660, re-annotated, re-trained. After all, the accuracy is not improved.
![2](https://user-images.githubusercontent.com/20389638/43444448-f1938a8a-94cd-11e8-98ee-cb57a58bcf0d.png)
**So, please recommend me what should I do to improve the accuracy in my case!**

**Comment by justinshaohi** [comment]
@AliceDinh du you has find some idea to improve?

**Comment by justinshaohi** [comment]
@AliceDinh i'm trying faster-rcnn-lowproposals that has high accuary now,it cost 30s in huawei mate10, what cost time you want be?

**Comment by AliceDinh** [comment]
@justinshaohi Faster R-CNN of course give high accuracy but I want an Android app, so Faster R-CNN is not suitable. What you mean by faster-rcnn-lowproposals?

**Comment by justinshaohi** [comment]
@AliceDinh <10s you can try ssd_resnet_fpn,about 8s. i use faster_rcnn_inception_resnet_v2_lowproposals

**Comment by AliceDinh** [comment]
@justinshaohi what's about the accuracy? my model cannot detect all 16 digits in cards

**Comment by justinshaohi** [comment]
@AliceDinh [ssd_fpn](https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/ssd_resnet50_v1_fpn_shared_box_predictor_640x640_coco14_sync.config) ,[faster_rcnn](https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/faster_rcnn_inception_resnet_v2_atrous_coco.config),you should modify [proposal](https://github.com/tensorflow/models/blob/d64bcfe3b56afd266887bb5e3553148e7b25eb5f/research/object_detection/samples/configs/faster_rcnn_inception_resnet_v2_atrous_coco.config#L45)

**Comment by AliceDinh** [comment]
@justinshaohi thanks, I'll let you know the result

**Comment by justinshaohi** [comment]
@AliceDinh hi,Are you making some progress？

**Comment by AliceDinh** [comment]
@justinshaohi not yet, I'll let you know soon

**Comment by derekjchow** [comment]
Could you try a different SSD model? [Resnet50 with FPN](https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/ssd_resnet50_v1_fpn_shared_box_predictor_640x640_coco14_sync.config) is likely a better choice if you want more accuracy.

**Comment by justinshaohi** [comment]
@derekjchow hi,what about your suggestion？Resnet50 with fpn in coco the ap is hightest。

**Comment by AliceDinh** [comment]
@justinshaohi training SSD Resnet50 v1 FPN shared box predictor  640x640 coco14 sync took me whole day to fix errors. It is running now but very slow and the training process is very strange to me (model_main.py), look so different from previous API version that trainned with train.py.

**Comment by AliceDinh** [comment]
@derekjchow  and @justinshaohi 
I trained SSD Resnet50 v1 FPN shared box predictor 640x640 using my data successfully. At the step 90, the loss = 0.8, I exported the model to frozen version and test on Jupyter Notebook but there is no object detected at all.

**Comment by justinshaohi** [comment]
@AliceDinh do you set fine_tune_checkpoint and from_detection_checkpoint is true,it can help you training faster.
and below is my loss ,about 2k step to 0.7
![_20180809144901](https://user-images.githubusercontent.com/16379061/43882754-783ec93e-9be3-11e8-9c7c-d9639bddf21b.jpg)

**Comment by AliceDinh** [comment]
@justinshaohi yes I did set this: fine_tune_checkpoint and from_detection_checkpoint is true. Maybe I need to wait for more steps to get the loss<1.0

**Comment by justinshaohi** [comment]
@lxyyang you don't need to freaze,just put the [frozen_inference_graph.pb](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/exporting_models.md) to asset,and modify [here](https://github.com/tensorflow/tensorflow/blob/f2494ede823da68abc4662460d8d951f63cb00e1/tensorflow/examples/android/src/org/tensorflow/demo/DetectorActivity.java#L66)

**Comment by AliceDinh** [comment]
Please tell me a bit about my training progress, coz I tested the frozen model and still no object detected.
And there are two graphs of loss, I am confused of  meaning of each. Please help.
![capture1](https://user-images.githubusercontent.com/20389638/43935977-aed52400-9c80-11e8-8332-88c8fdd8da9b.PNG)
![capture2](https://user-images.githubusercontent.com/20389638/43935978-af0df2b2-9c80-11e8-87c5-b62b810ad3db.PNG)
![capture3](https://user-images.githubusercontent.com/20389638/43935980-af4b7ec0-9c80-11e8-9c4f-3c9c745b6db2.PNG)
![capture4](https://user-images.githubusercontent.com/20389638/43935981-af7eccee-9c80-11e8-9a40-5274db26a3c1.PNG)
![capture](https://user-images.githubusercontent.com/20389638/43935982-afb3f2a2-9c80-11e8-914d-2633de9f3ad5.PNG)

**Comment by AliceDinh** [comment]
@justinshaohi and @derekjchow 
ssd_mobilenet_v1_fpn gives higher accuracy, that's very true. I trained the model and when the total loss is arround 0.5, it detected all objects as my expectation on Jupyter Notebook. That's great, so I imported to [TensorFlow Android Camera Demo](https://github.com/tensorflow/tensorflow/tree/r1.10/tensorflow/examples/android) at DetectorActivity.java, I modified 
` private static final String TF_OD_API_MODEL_FILE =
      "file:///android_asset/frozen_inference_graph.pb";
private static final String TF_OD_API_LABELS_FILE = "file:///android_asset/my_list.txt";`
But, nothing detected. Just to confirm that the app was working with the model I trained from SSD mobilenet V1

**Comment by justinshaohi** [comment]
@AliceDinh modify image size to 640

**Comment by AliceDinh** [comment]
@justinshaohi 
You mean modifying 
`private static final int TF_OD_API_INPUT_SIZE = 300;` 
to 
`private static final int TF_OD_API_INPUT_SIZE = 640;`
I did but nothing change

**Comment by AliceDinh** [comment]
@justinshaohi I don't understand, can you tell me more details?

**Comment by justinshaohi** [comment]
@AliceDinh set true,the image  will save in sdcard,go to sdcard see the image is have the object that you want detection

**Comment by AliceDinh** [comment]
@justinshaohi no image is saved into SD card

**Comment by AliceDinh** [comment]
@justinshaohi there is only one image saved into SD card, named preview.png but I cannot view it.

**Comment by justinshaohi** [comment]
@AliceDinh ssd_fpn compute need cost 10s，so you should open camera aim at object for 10s，or you use recognition once not real time

**Comment by AliceDinh** [comment]
@justinshaohi still nothing detected, I did hold the camera for a long time

**Comment by derekjchow** [comment]
@seanpmorgan It just hasn't been implemented yet. I think it's on the roadmap.

**Comment by maxcrous** [comment]
@AliceDinh and @lxyyang or anyone else, have you been able to deploy ssd_mobilenet_v1_fpn to mobile? 
When I train ssd_mobilenet_v1_fpn 

- I reach a low loss. 
- I am able to export_tflite_ssd_graph succesfully
- I am able to tflite_convert succesfully

But, when loading the model I receive a Fatal signal 6 (SIGABRT). 

The same happens when using the frozen_inference_graph supplied with the download from the model zoo.

**Comment by AliceDinh** [comment]
@maxcrous 
Can you create a new issue on that, because I was not able to deploy ssd-fpn frozen model to mobile as well.

**Comment by tesfagabir** [comment]
@AliceDinh were you able to convert Faster RCNN frozen graph (pb) to tflite format. I am having issues in converting it.

**Comment by lucassardelari** [comment]
Hi, I have the TF demo project, and I have a neural network trained with faster rcnn inception v2 coconut and I would like to run it on android to perform reconnaissance.
I'm having an exception when Scale them back to the input size.
TensorFlowObjectDetectionAPIModel.java File
Line: 193

I already did the change:
private static final int MAX_RESULTS = 500;

Someone could tell me if I already had this exception or if I was able to use this neural network with Android.

**Comment by thusinh1969** [comment]
@AliceDinh In Android apps (Android studio ?), set TF_OD_..._QUANTIZE to False.

**Comment by ahamino** [comment]
> @seanpmorgan It just hasn't been implemented yet. I think it's on the roadmap.

@derekjchow You mentioned earlier that crop-and-resize step is on the roadmap for TFLite. Any update on when it is expected this operation will be added to TFLite?

**Comment by ajinkya933** [comment]
@AliceDinh try reducing number of classes to1 (text). and then use mobile net.

**Comment by hadizand** [comment]
> @AliceDinh were you able to convert Faster RCNN frozen graph (pb) to tflite format. I am having issues in converting it.

Could you find any solution for converting frcnn to TFlite?

**Comment by seanpmorgan** [comment]
> > @AliceDinh were you able to convert Faster RCNN frozen graph (pb) to tflite format. I am having issues in converting it.
> 
> Could you find any solution for converting frcnn to TFlite?

See the recent announcement for MLIR based TFLite. It supports control flow and architectures like Faster RCNN:

https://groups.google.com/a/tensorflow.org/forum/#!topic/tflite/C7Ag0sUrLYg
